active/0051-strategies.md

```diff
@@ -0,0 +1,368 @@
+  graph](https://en.wikipedia.org/wiki/Transitive_reduction#Graph_algorithms_for_transitive_reduction));
++ producing an application tree with only used modules
+
+Dependency graph represents dependencies using module names, there is a need to
+be able to convert module name to file path. This is where `File System
+Resolver` comes in. Here's couple of examples:
+
+```javascript
+fileSystemResolver.resolve('lodash') => `some-path/node_modules/lodash/lodash.js`
+fileSystemResolver.resolve('ember-ajax') => `some-path/addon-modules/ember-ajax/index.js`
+fileSystemResolver.resolve('ember-data') => `some-path/addon-modules/modules/ember-data/index.js`
+fileSystemResolver.resolve('ember-data/-private') => `some-path/addon-modules/modules/-private.js`
+```
+
+This effort could be broken down into several phases:
+
++ dead modules elimination inside of the `addons/` (application would be the
+  main entry point and unused modules are removed only from `addons/`)
++ dead modules elimination inside of the `app/`
+  + removing unused components and helpers (requires analysing templates)
+  + removing unused initializers/services (this likely entails work on
+    dependency injection layer as we would need access to a resolver resolution
+    map)
++ treeshaking (Rollup-like treeshaking where we include *only* the code that is
+  used)
+
+`Linker` would be able to take an `exclude` list of modules as a parameter.
+Although, valuable in some situations, it should be clearly marked as advanced
+API. It should be used as a last resort and serve as an "escape hatch".
+
+It would make sense to implement `Linker` as a strategy. Developers would be
+able to "opt in"/"opt out" of optimising behaviour.
+
+## Deprecating `app.import` API
+
+Ember applications which choose to use `Linker` strategy should be able to
+remove usages of `app.import`.
+
+## Tools
+
+With growing complexity of Ember applications, it is crucial to provide more
+insights into final assets.
+
+Main goals are:
+
+- report raw/uglified/compressed asset sizes;
+  [broccoli-concat-analyser](https://github.com/stefanpenner/broccoli-concat-analyser)
+- find source code duplication across your javascript assets (enables you to
+  fine tune code splitting parameters to reduce bundle invalidation rates as
+  well as improve repeat page load performance)
+
+# How We Teach This
+
+This is a backward compatible change to the existing Ember CLI eco system. In
+order to teach users how to use strategies, we need to update the API docs with
+a section for this and the best practices of when to use this. A more general
+purpose blog post could be beneficial as well.
+
+# Drawbacks
+
+There are several potential drawbacks that are worth noting.
+
+_Build performance_. There is minimal overhead in instantiating strategies and
+calling methods on them and I believe this approach shouldn't degrade build
+performance.
+
+_Coupling with Broccoli_. The idea of having a more generic cli was discussed
```

Comment:
This is probably not worth mentioning IMO. The discussions are making it generic were Ember CLI w/ broccoli + angular/react/vue etc. This RFC wouldn't affect any of those considerations.